keras/callbacks/swap_ema_weights.py

```diff
@@ -0,0 +1,77 @@
+from keras import ops
+from keras.api_export import keras_export
+from keras.callbacks.callback import Callback
+
+
+@keras_export("keras.callbacks.SwapEMAWeights")
+class SwapEMAWeights(Callback):
+    """Swaps EMA weights before and after the evaluation.
+
+    `SwapEMAWeights` callback is used in conjunction with the optimizer using
+    `use_ema=True`.
+
+    Example:
+
+    ```python
+    # Remember to set `use_ema=True`
+    model.compile(optimizer=SGD(use_ema=True), loss=..., metrics=...)
+
+    # Metrics will be computed with EMA weights
+    model.fit(X_train, Y_train, callbacks=[SwapEMAWeights()])
+
+    # If you want to save model checkpoint with EMA weights, you can set
+    # `swap_on_epoch=True` and place ModelCheckpoint before SwapEMAWeights.
+    model.fit(
+        X_train,
+        Y_train,
+        callbacks=[ModelCheckpoint(...), SwapEMAWeights(swap_on_epoch=True)]
+    )
+    ```
+
+    Args:
+        swap_on_epoch: whether to perform swapping `on_epoch_begin` and
+            `on_epoch_end`. This is useful if you want to use EMA weights for
+            other callbacks such as `ModelCheckpoint`. Defaults to `False`.
+
+    """
+
+    def __init__(self, swap_on_epoch=False):
+        super().__init__()
+        self.swap_on_epoch = swap_on_epoch
+
+    def _swap_variables(self):
```

Comment:
I find the "swap" system a bit confusing when it comes to tracking what is now what. It may be better to have "set model weights to ema" and "revert model weights". But I'm guessing you're reusing the EMA variables for weight storage to save memory?

Reply:
Yes. It could significantly increase memory usage to create an additional copy of the model weights, especially for large models.